packages/loader/container-loader/src/container.ts

```diff
@@ -810,6 +818,41 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         }
     }
 
+    private _disposed = false;
+    private disposeCore(error?: ICriticalContainerError) {
+        assert(!this._disposed, "Container already disposed");
+        this._disposed = true;
+
+        try {
+            // Ensure that we raise all key events even if one of these throws
+            try {
+                this._protocolHandler?.close();
+
+                this.connectionStateHandler.dispose();
+
+                this._context?.dispose(error !== undefined ? new Error(error.message) : undefined);
+
+                this.storageService.dispose();
```

Comment:
I know this is merged, but I think this is not doing what the comment says we should? If `this._protocolHandler?.close()` throws, the other objects won't be disposed and will never be because the method prevents a second call through the `assert()` at the top. This feels like it should do independent try catches for each sub-object that needs to be disposed.

Reply:
Hm fair point, but the same thing was happening before:
1. Set lifecycle state to "closing"
2. Error could occur during any of the separate close/dispose calls
3. Catch and log error
4. Set lifecycle state to "closed" (which would prevent re-entering the "closeCore" flow)

Also, since "closeCore" and "disposeCore" currently get called as responses to DeltaManager events, we won't ever reach that case you described unless we try to call those methods in other places.

Reply:
Yeah, I noticed that. Which I found a bit weird tbh 😄 , mostly because of the potential for the container to still get called after it kicks off DeltaManager's closing, but before it hears back from it, even though `container.close()/dispose()` was already called and returned. Something to keep in mind, maybe.